File: ios/Classes/SessionInfoProvider.swift
import Metrix
import Flutter

class SessionIdStreamHandler: NSObject, FlutterStreamHandler {
    private var sessionIdEventSink: FlutterEventSink? = nil

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sessionIdEventSink = events
        MetrixClient.setSessionIdListener { (sessionId: String) in
            self.sessionIdEventSink?(sessionId)
        }
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sessionIdEventSink = nil
        return nil
    }
}

class SessionNumberStreamHandler: NSObject, FlutterStreamHandler {
    private var sessionNumberEventSink: FlutterEventSink? = nil
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sessionNumberEventSink = events
        MetrixClient.setSessionNumberListener { (sessionNum: Int) in
            self.sessionNumberEventSink?(sessionNum)
        }
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sessionNumberEventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/ir/metrix/flutter/SessionInfoProvider.kt
package ir.metrix.flutter

import io.flutter.plugin.common.EventChannel
import ir.metrix.analytics.SessionIdListener
import ir.metrix.analytics.SessionNumberListener

class SessionIdReader: EventChannel.StreamHandler {
    private var sessionIdEventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sessionIdEventSink = events
        ir.metrix.analytics.MetrixAnalytics.setSessionIdListener(object : SessionIdListener {
            override fun onSessionIdChanged(sessionId: String) {
                sessionIdEventSink?.success(sessionId)
            }
        })
    }

    override fun onCancel(arguments: Any?) {
        sessionIdEventSink = null
    }
}

class SessionNumReader: EventChannel.StreamHandler {
    private var sessionNumEventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sessionNumEventSink = events
        ir.metrix.analytics.MetrixAnalytics.setSessionNumberListener(
            object : SessionNumberListener {
                override fun onSessionNumberChanged(sessionNumber: Int) {
                    sessionNumEventSink?.success(sessionNumber)
                }
            })
    }

    override fun onCancel(arguments: Any?) {
        sessionNumEventSink = null
    }
}